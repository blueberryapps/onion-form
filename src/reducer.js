import * as actions from './actions';
import { Record, Map } from 'immutable';

export const InitialState = Record({
  fields: new Map(),
  forms: new Map()
});

const initialState = new InitialState;

export const deafultFieldProperties = Map({
  value: '',
  liveValidation: false,
  error: null,
  apiError: null
});

function reviveFields(fields) {
  return Object.keys(fields).reduce(
    (acc, field) => acc.set(field, deafultFieldProperties.merge(fields[field])),
    new Map
  );
}

function revive(state) {
  const fields = (state.fields || {});
  return Object.keys(fields).reduce(
    (acc, form) => acc.setIn(['fields', form], reviveFields(fields[form]))
    , initialState
  );
}

/**
 * Translation Redux reducer
 * @param  {Object} inputState  App state
 * @param  {Object} action      Flux action
 * @return {Object}             Updated app state
 */
export default function translationReducer(inputState = initialState, action = {}) {
  const state = !(inputState instanceof InitialState) ? revive(inputState) : inputState;

  switch (action.type) {
    case actions.REGISTER_ONION_FORM_FIELD: {
      const { form, field } = action;
      if (Map.isMap(state.getIn(['fields', form, field]))) {
        return state;
      }
      return state.setIn(['fields', form, field], deafultFieldProperties);
    }

    case actions.SET_ONION_FORM_FIELD_PROPERTY: {
      const { form, field, property, value } = action;
      return state.setIn(['fields', form, field, property], value);
    }

    case actions.SET_ONION_FORM_MULTIPLE_FIELDS: {
      const { form, property, values } = action;
      return Object.keys(values).reduce(
        (acc, field) => acc.setIn(['fields', form, field, property], values[field]),
        state
      );
    }

    case actions.CLEAR_ONION_FORM: {
      const { form } = action;
      const fields = state.getIn(['fields', form]) || [];
      return fields.reduce(
        (acc, _, field) => acc.setIn(['fields', form, field], deafultFieldProperties),
        state
      );
    }

    case actions.CLEAR_ONION_FORM_PROPERTY: {
      const { form, property } = action;
      const fields = state.getIn(['fields', form]) || [];
      return fields.reduce(
        (acc, _, field) => acc.setIn(['fields', form, field, property], null),
        state
      );
    }
  }

  return state;
}
