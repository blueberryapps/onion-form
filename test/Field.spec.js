/* eslint-disable react/no-multi-comp */

import Field, { BasicInput } from '../src/Field.react';
import Form from '../src/Form.react';
import React from 'react';
import reducer from '../src/reducer';
import TestUtils from 'react-addons-test-utils';
import { assert } from 'chai';
import { createStore } from 'redux';
import { jsdom } from 'jsdom';
import { Provider as ReduxProvider } from 'react-redux';

const initial = {
  fields: {
    fooForm: {
      firstName: {
        customProperty: 'Hi Hello From State',
        value: 'Bar',
        error: 'isRequired'
      }
    }
  }
};

const store = createStore(() => ({ onionForm: reducer(initial, {}) }));

global.document = jsdom('<!doctype html><html><body></body></html>');
global.window = document.defaultView;

describe('Field', () => {
  function createStub(customProps = {}, Input) {
    const container = TestUtils.renderIntoDocument(
      <ReduxProvider store={store}>
        <Form name="fooForm">
          <Field name="firstName" {...customProps} />
        </Form>
      </ReduxProvider>
    );

    return TestUtils.findRenderedComponentWithType(container, Input);
  }

  const stub = createStub({}, BasicInput);

  it('input should have name prop', () => {
    assert.equal(stub.props.name, 'firstName');
  });

  it('input should have onChange prop', () => {
    assert.typeOf(stub.props.onChange, 'function');
  });

  it('input should have value prop', () => {
    assert.equal(stub.props.value, 'Bar');
  });

  it('input should have error prop', () => {
    assert.equal(stub.props.error, 'isRequired');
  });

  it('input should have customProperty prop', () => {
    assert.equal(stub.props.customProperty, 'Hi Hello From State');
  });

  describe('with custom Component', () => {
    const FooInput = () => (<div className="FooInput" />);

    it('should render it', () => {
      const fooContainer = TestUtils.renderIntoDocument(
        <ReduxProvider store={store}>
          <Form name="fooForm">
            <Field name="firstName" component={FooInput} />
            <FooInput />
          </Form>
        </ReduxProvider>
      );

      TestUtils.scryRenderedDOMComponentsWithClass(fooContainer, 'FooInput');
    });
  });
});
