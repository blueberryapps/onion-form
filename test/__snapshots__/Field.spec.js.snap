// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field with custom Component should render it 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Form
    method="POST"
    name="fooForm"
    validations={Object {}}
  >
    <form
      method="POST"
      onSubmit={[Function]}
    >
      <Field
        component={[Function]}
        name="firstName"
      >
        <FormfirstNameContextExtractor>
          <FormfirstNameField
            name="firstName"
            onionFieldRegister={[Function]}
            onionFormName="fooForm"
            onionIsValid={[Function]}
            onionLiveValidate={[Function]}
            onionOnSubmit={[Function]}
          >
            <Field
              actions={
                Object {
                  "clearForm": [Function],
                  "clearFormProperty": [Function],
                  "registerField": [Function],
                  "setFieldApiError": [Function],
                  "setFieldError": [Function],
                  "setFieldLiveValidation": [Function],
                  "setFieldValue": [Function],
                  "setFormFieldProperty": [Function],
                  "setMultipleFields": [Function],
                }
              }
              field={
                Immutable.Map {
                  "value": "Bar",
                  "liveValidation": false,
                  "error": "isRequired",
                  "apiError": null,
                  "customProperty": "Hi Hello From State",
                }
              }
              name="firstName"
              onionFieldRegister={[Function]}
              onionFormName="fooForm"
              onionIsValid={[Function]}
              onionLiveValidate={[Function]}
              onionOnSubmit={[Function]}
            >
              <FooInput
                actions={
                  Object {
                    "clearForm": [Function],
                    "clearFormProperty": [Function],
                    "registerField": [Function],
                    "setFieldApiError": [Function],
                    "setFieldError": [Function],
                    "setFieldLiveValidation": [Function],
                    "setFieldValue": [Function],
                    "setFormFieldProperty": [Function],
                    "setMultipleFields": [Function],
                  }
                }
                apiError={null}
                component={[Function]}
                customProperty="Hi Hello From State"
                error="isRequired"
                field={
                  Immutable.Map {
                    "value": "Bar",
                    "liveValidation": false,
                    "error": "isRequired",
                    "apiError": null,
                    "customProperty": "Hi Hello From State",
                  }
                }
                hint={null}
                label={null}
                liveValidation={false}
                name="firstName"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onionFieldRegister={[Function]}
                onionFormName="fooForm"
                onionIsValid={[Function]}
                onionLiveValidate={[Function]}
                onionOnSubmit={[Function]}
                placeholder={null}
                tooltip={null}
                value="Bar"
              >
                <div
                  className="FooInput"
                />
              </FooInput>
            </Field>
          </FormfirstNameField>
        </FormfirstNameContextExtractor>
      </Field>
      <FooInput>
        <div
          className="FooInput"
        />
      </FooInput>
    </form>
  </Form>
</Provider>
`;
